Ecto 3 ongoing work

joeapearson opened the pull request:
Changes so far:

* Fixed `binary_id` test.  It's difficult to understand what the purpose of `Ecto.Query.Tagged` since there is little documentation available so, while the test now passes, I'm not certain if the change I've made will stand the test of time.
* Lots of cleanup, fixed warnings, added github actions for validation

Again... I've been quite ruthless in not attempting any form of backwards compatibility.  That can come later (or we can cut this as a major version if we ever get to the stage where this can be published).

joeapearson committed:
Add github actions replacing travis

joeapearson committed:
Remove superfluous config

joeapearson committed:
Deprecate ebert

joeapearson committed:
Add credo and dialyzer

joeapearson committed:
Add ongoing notes to cover support of ecto3

joeapearson committed:
Update shields

joeapearson committed:
Cleanup disused deps

joeapearson committed:
Credo / formatting fixes

joeapearson committed:
Check formatting in CI

joeapearson committed:
Fix binary_id test

joeapearson committed:
Use DateTime.from_naive API

joeapearson committed:
Add index mgmt APIs

justinmorrow commented:
@joeapearson Found a little insight into `Ecto.Query.Tagged` from their git log. I have to admit I was completely lost about what the hell `Tagged` means too haha

> commit b9c6c79973c1220c0ffdb4dca9e1522c6b019284
> Author: Martin Schürrer <martin@schuerrer.org>
> Date:   Fri Jul 25 13:58:28 2014 -0500
> 
>     Unify Ecto.Array and Ecto.Binary into Ecto.Tagged. Fixes #230.

The discussion in https://github.com/elixir-ecto/ecto/pull/230 explains how it's meant to be used a bit. Essentially it's a way to define a parameterized array type for arrays containing binaries, which wasn't really supported before apparently.

Really appreciate you sharing the work you're doing here by the way! Feel free to @mention me when it's ready for review (though I can't approve it), or if you're stuck on anything. I'm still assessing whether or not the ecto 3 upgrade makes sense for ~~our~~ _my particular_ project, but if it does I'd be happy to contribute to this and help move things along. Glad to help with smallish things either way anyways, whenever I have the time for it :)

joeapearson committed:
First fixed upsert test

joeapearson commented:
Today added a first passing `on_conflict` (upsert) test.  @scottmessinger  FYI!

It's a complete mess still but anyone is welcome to review and add comments.  Don't expect anything to work yet™️.  But you might wish to comment now if you see the changes I'm making going in an undesirable direction.

In order to accomplish this I was forced to add support for multiple updates in a single query to the mongodb driver.  Those changes can be found at https://github.com/avid-technology/mongodb/tree/ecto-3

Next things on my list are:

* Continue smashing through the `on_conflict` test cases.  I'm not certain whether it's going to be possible to support everything yet.
* Cleanup

scottmessinger commented:
@joeapearson This is really exciting! I've long wanted upserts to be added to the driver. I'm pumped to be able to roll these out!

joeapearson committed:
Allow dev-only .iex.exs file

joeapearson committed:
All upsert tests pass

joeapearson committed:
Add requirement for mongo 4.2

joeapearson commented:
Update:

1. All `:with_conflict_target` tests now pass (with caveats, below).
2. Some `:without_conflict_target` pass too.

Here are the caveats:

1. Primary ID is no longer automatically generated and is instead generated by the driver.  The Ecto upsert behaviour implies this need because presence or lack of returned `id` field is used to determine whether an insert actually took place in the event of conflict.  This in turn has created a need for supporting `:read_after_writes` at least in the case of `id`.  I think it might be possible to support `:read_after_writes` for all fields but I'm not focusing on this now.
2. As far as I can tell Mongo will **never** be able to support the `on_conflict: :replace_all` option, because Mongo does not allow replacement of the `_id` field.  The only way of doing this is a delete followed by an insert in two separate operations, so data loss is possible.  I propose to allow this but print a huge error warning to the console that developers can optionally disable if they understand the risk.  Allowing this will make the tests pass but will also make it easier for a would be adoptee to use existing code without having to modify it.
3. The `Custom` ecto schema (from the Ecto integration tests) specify a primary key of `bid` as opposed to the usual `id`.  I suspect this can never be supported, but possibly we could enforce the mapping of primary key to `_id` in all cases.


**Biggest caveat:**  the code still continues to be a total mess.  My next task is to clean this up.

joeapearson committed:
Insert single docs refactored

joeapearson committed:
insert functions refactored

joeapearson committed:
Factor projection into command fn

joeapearson committed:
Cleanup

joeapearson committed:
Cleanup

joeapearson committed:
Fix non id primary key tests

joeapearson committed:
Refactored insert_all, all tests pass

joeapearson committed:
Correct impl of conflict_target

joeapearson committed:
Cleanup for beta upsert support

joeapearson commented:
Code is in a better state.  I'd say worthy of "alpha" and some feedback.  I don't have a great deal of time to work on this for a while now.

The news is not entirely fantastic.  I don't think we should recommend users attempt upserts in anything other than simple cases.  I also had to make some non-backwards compatible changes to the `mongodb` driver itself in cases where the driver wasn't returning enough information back to `mongodb_ecto` in order to accommodate the required upsert functionality.  Some upsert operations require Mongo update pipelines which only appeared in MongoDB 4.2.

What I've discovered is that MongoDB support for upserts is just not as rich as it is in other databases.  There are a couple of spots where I've added a `raise` to indicate either an unsafe upsert (involving multiple DB calls) or a completely unsupported feature.

Supported but require multiple db calls and therefore unsafe:

* Any `on_conflict` that involves replacing the `id` (`_id`) field, e.g. `replace_all`.  The `_id` field is immutable in Mongo so the only way around it is to first delete any matching doc and then attempting performing the insert.  Not nice.

The recommended approach is to avoid these unsafe calls but optionally a developer can pass an option of `allow_unsafe_upserts: true` to allow it.  That's how the tests work.

Completely not supported:

* Passing a query to on_conflict with a subsequent write operation, e.g. `on_conflict: [set: [field: "xyz"]]`.  This could probably be achieved with multiple DB calls but I made the decision that it was just getting ridiculous.  If anyone disagrees then I suggest we treat this as a PR for the future.

Writing the code was a process of discovery for me both in terms of Ecto and the mongodb libraries.  I have no doubt that improvements could be made.  All feedback is welcome!

Lastly, in order to try this out you'll have to use the updates I've added at https://github.com/avid-technology/mongodb/tree/ecto-3  .

joeapearson committed:
Remove disused find_one_and_replace

joeapearson committed:
Remove test file

scottmessinger commented:
Just wanted to follow up with this:

> Again... I've been quite ruthless in not attempting any form of backwards compatibility. 

Is this still true? And, if so, what area are not backwards compatible?

scottmessinger commented on lib/mongo_ecto/connection.ex:
```diff
@@ -307,7 +373,77 @@ defmodule Mongo.Ecto.Connection do
+          To workaround this issue you may:
+
+          * Use a different `on_conflict` strategy (this is the safest option)
+          * If you must use `:replace_all`, you may pass an additional `allow_unsafe_upserts: true` option.
```
I think this is a good requirement. nice!

scottmessinger commented:
> Some upsert operations require Mongo update pipelines which only appeared in MongoDB 4.2.

Would there be a way to have someone opt into this with a flag or by us knowing what version on Mongo they're using? I'm just trying to think through how we make sure people don't use a feature that blows up on 4.0 and they're quite confused as to why.

joeapearson commented:
> Is this still true? And, if so, what area are not backwards compatible?

Yes still true.  The changes to the `mongodb` `findAndModify` related calls are non-backwards compatible.  Upserts that require MongoDB update pipelines won't work on Mongo < 4.2. 

As far as I can see the `findAndModify` implementation in `mongodb` was oversimple and obscured some useful info coming back from Mongo.  Maybe we could introduce a "legacy behaviour" option for users that still want it

As for Mongo 4.2, we probably could figure out what version of Mongo we are using but, well... Mongo 4.0 is now 3 odd years old.  If I find time I'll take a look and see if we can figure it out.

Lastly, I made some changes to support OTP 24.  I didn't make those backwards compatible either.

joeapearson committed:
Point to supported mongodb driver
